Simplified Update Messages

darth-cheney opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
This PR refactors Messenger so that it sends simplified messages to the frontend

## Motivation and Context
Until now, we would send over entire nested, recursive subtrees of any updated and/or created node during a single recalculate cycle. This is fine for small subtrees, but when the affected nodes are high in the overall structure, it means we are -- unnecessarily -- sending over a lot of extra data with each update. By extra unnecessary data here we mean serialized descriptions of Cell nodes that haven't changed at all and therefore we really don't need to send them.
<!-- Why is this change required? -->
  
Simplifying this has the following advantages:
1. On the Cells side, we do far fewer recursive descents in order to serialize extra nodes. Now we only fully serialize nodes that have actually changed.
2. There is far less data going over the websocket, aside from the initial total renders of the page
3. Less work is required by the CellHandler when dealing with update messages. For unchanged children, we simply look them up by ID, put them in order, and there's nothing else to do.
<!-- What problem does it solve? -->
<!--  If it fixes an open issue, please link to the issue here.-->

## Approach
<!-- Describe your changes in reasonable detail -->
Because we have a better idea of when a given node `wasCreated` or `wasUpdated` during a given recalc cycle, we can leverage this knowledge when composing messages that go over the wire.
  
In the case of any node that was just created, we need to serialize the whole thing -- including all of its children and children's children, recursively. This is equivalent to what we were doing already.
  
The difference comes with nodes that were only updated. In this case, we do the following:
1. Create a smaller base serial description of the updated node (id, children, etc)
2. For each of the node's children, we check if the node was created, updated, or if nothing happend
    2.1. For a newly created child , serialize a full description recursively
    2.2. For a child that was also updated, go to Step 1
    2.3. For a child that was not updated, not created, and not discarded, simply return its identity
  
The savings really comes from step 2.3 (and 2.2 which precedes it).

## How Has This Been Tested?
<!-- Describe in reasonable detail how you tested your changes. -->
All Cells and JS tests are passing.
  
Changes were tested in the demo frontend and there were no errors after poking around. But I recommend testing on the live production system before merging.
<!-- If appropriate, include details of your testing environment, -->
<!-- tests ran to see how your change affects other areas of the code, etc. -->

## Screenshots or console output (if appropriate)
<!-- if not appropriate, remove this section -->

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

darth-cheney committed:
Adding simplified Messages (id only) from Python and handling for them web side

braxtonmckee commented on object_database/web/cells/Messenger.py:
```diff
@@ -191,3 +187,51 @@ def _resolveExpandedChild(parent_id, cell_or_list, name_in_parent):
     return _getExpandedStructure(parent_id, cell_or_list, name_in_parent)
+
+
+"""New shit down here"""
```
?

braxtonmckee reviewed (commented):
stuff doesn't appear to be updating correctly - for instance I click a file in the research frontend to open it with this change, and the selection doesn't update.

darth-cheney committed:
Adding demos for Plot

darth-cheney committed:
Fixing bug where #componentDidUpdate not triggered on Subscribed contents

This came as the result of the new "abbreviated" update message structure,
where we only pass IDs for certain children that have not changed at all.

The solution implemented here was to leverage the `.isSubscribed` property
of Subscribed components in the handler, and to force-add their contents
to the list of "updated components," who will then have #componentDidUpdate
called.

I have added tests for this and related scenarios.

darth-cheney committed:
Updating base Component to better deal with Subscribed wrapping.

This commit also ensures that the contents of any Subscribed that
was updated by ID alone will recursively be added (with children)
to the update list to have #componentDidUpdate called. This was
not happening before.

darth-cheney committed:
Updating lifecycle handling for created and updated Components.

Previously, CellHandler kept internal lists of updated and
created components, and would push to them where needed. These
are used at the end of a cycle to call #componentDidLoad (created)
and #componentDidUpdate (updated) respectively.

It turns out we could find ourselves pushing the same component
to this list multiple times, triggering multiple lifecycle method
calls in a single cycle. This led to unexpected behavior.

Instead of changing the lists to a Set, we now have properties
on the base Component prototype that have default settings. We
collect/select for these attributes once each cycle, call the
appropriate lifecycle method, and then reset the property to false.

darth-cheney committed:
Revert "Updating base Component to better deal with Subscribed wrapping."

This reverts commit ee88f6422db0800ec8c20c2698b3bc77a0a32e5b.

darth-cheney committed:
Fixing SubscribedSequence error where removals went unsent.

When elements were removed from a nested SubscribedSequence,
an updated message was not sent in certain cases. This was due
to the use of rootMergeNode() and not setting its wasUpdated property
to True when in a recalculation cycle.

darth-cheney committed:
Re-Adding old_sheet cells demos

darth-cheney committed:
Updating comments
